File: cs/Basic/Server.NodeTypes/Nodes/MyComplexVariableNode.cs
// Copyright (c) Traeger Industry Components GmbH. All Rights Reserved.

namespace NodeTypes
{
    using System;
    using Opc.UaFx;

    internal class MyComplexVariableNode : OpcDataVariableNode<int>
    {
        #region ---------- Private fields ----------

        private OpcDataVariableNode<string> mandatoryValueNode;
        private OpcDataVariableNode<int> optionalValueNode;

        #endregion

        #region ---------- Public constructors ----------

        public MyComplexVariableNode(
                OpcName name,
                string mandatoryValue,
                int optionalValue = -1)
            : base(name)
        {
            this.InitializeChildren(mandatoryValue, optionalValue);
        }

        public MyComplexVariableNode(
                IOpcNode parent,
                OpcName name,
                string mandatoryValue,
                int optionalValue = -1)
            : base(parent, name)
        {
            this.InitializeChildren(mandatoryValue, optionalValue);
        }

        #endregion

        #region ---------- Public properties ----------

        public string MandatoryValue
        {
            get => this.mandatoryValueNode.Value;
            set => this.mandatoryValueNode.Value = value;
        }

        public bool HasOptionalValue
        {
            get => this.optionalValueNode != null;
        }

        public int OptionalValue
        {
            get
            {
                if (this.optionalValueNode == null)
                    throw new NotSupportedException();

                return this.optionalValueNode.Value;
            }

            set
            {
                if (this.optionalValueNode == null)
                    throw new NotSupportedException();

                this.optionalValueNode.Value = value;
            }
        }

        #endregion

        #region ---------- Protected properties ----------

        protected override OpcNodeId DefaultTypeDefinitionId
        {
            get => "ns=1;s=MyComplexVariableType";
        }

        #endregion

        #region ---------- Private methods ----------

        private void InitializeChildren(string mandatoryValue, int optionalValue)
        {
            this.mandatoryValueNode = new OpcDataVariableNode<string>(
                    parent: this,
                    name: "1:MandatoryValue",
                    value: mandatoryValue);

            if (optionalValue != -1) {
                this.optionalValueNode = new OpcDataVariableNode<int>(
                        parent: this,
                        name: "1:OptionalValue",
                        value: optionalValue);
            }
        }

        #endregion
    }
}

File: vb/Basic/Server.NodeTypes/Nodes/MyComplexVariableNode.vb
' Copyright (c) Traeger Industry Components GmbH. All Rights Reserved.

Imports System
Imports Opc.UaFx

Namespace NodeTypes
    Friend Class MyComplexVariableNode
        Inherits OpcDataVariableNode(Of Integer)

#Region "---------- Private fields ----------"

        Private _mandatoryValueNode As OpcDataVariableNode(Of String)
        Private _optionalValueNode As OpcDataVariableNode(Of Integer)

#End Region

#Region "---------- Public constructors ----------"

        Public Sub New(
                ByVal name As OpcName,
                ByVal mandatoryValue As String,
                ByVal Optional optionalValue As Integer = -1)
            MyBase.New(name)
            Me.InitializeChildren(mandatoryValue, optionalValue)
        End Sub

        Public Sub New(
                ByVal parent As IOpcNode,
                ByVal name As OpcName,
                ByVal mandatoryValue As String,
                ByVal Optional optionalValue As Integer = -1)
            MyBase.New(parent, name)
            Me.InitializeChildren(mandatoryValue, optionalValue)
        End Sub

#End Region

#Region "---------- Public properties ----------"

        Public Property MandatoryValue As String
            Get
                Return Me._mandatoryValueNode.Value
            End Get
            Set(ByVal value As String)
                Me._mandatoryValueNode.Value = value
            End Set
        End Property

        Public ReadOnly Property HasOptionalValue As Boolean
            Get
                Return Me._optionalValueNode IsNot Nothing
            End Get
        End Property

        Public Property OptionalValue As Integer
            Get
                If Me._optionalValueNode Is Nothing Then Throw New NotSupportedException()
                Return Me._optionalValueNode.Value
            End Get

            Set(ByVal value As Integer)
                If Me._optionalValueNode Is Nothing Then Throw New NotSupportedException()
                Me._optionalValueNode.Value = value
            End Set
        End Property

#End Region

#Region "---------- Protected properties ----------"

        Protected Overrides ReadOnly Property DefaultTypeDefinitionId As OpcNodeId
            Get
                Return "ns=1;s=MyComplexVariableType"
            End Get
        End Property

#End Region

#Region "---------- Private methods ----------"

        Private Sub InitializeChildren(ByVal mandatoryValue As String, ByVal optionalValue As Integer)
            Me._mandatoryValueNode = New OpcDataVariableNode(Of String)(
                    parent:=Me,
                    name:="1:MandatoryValue",
                    value:=mandatoryValue)

            If optionalValue <> -1 Then
                Me._optionalValueNode = New OpcDataVariableNode(Of Integer)(
                        parent:=Me,
                        name:="1:OptionalValue",
                        value:=optionalValue)
            End If
        End Sub

#End Region
    End Class
End Namespace
